libsqlite3-sys/src/wasm32_unknown_unknown.rs

```diff
@@ -0,0 +1,181 @@
+    8, // wasm32 max_align_t
+    max(std::mem::size_of::<usize>(), std::mem::align_of::<usize>()),
+);
+
+#[no_mangle]
+pub unsafe extern "C" fn malloc(size: usize) -> *mut u8 {
+    let layout = match std::alloc::Layout::from_size_align(size + ALIGN, ALIGN) {
+        Ok(layout) => layout,
+        Err(_) => return null_mut(),
+    };
+
+    let ptr = std::alloc::alloc(layout);
+    if ptr.is_null() {
+        return null_mut();
+    }
+
+    *(ptr as *mut usize) = size;
+    ptr.offset(ALIGN as isize)
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn free(ptr: *mut u8) {
+    let ptr = ptr.offset(-(ALIGN as isize));
+    let size = *(ptr as *mut usize);
+    let layout = std::alloc::Layout::from_size_align_unchecked(size + ALIGN, ALIGN);
+
+    std::alloc::dealloc(ptr, layout);
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn realloc(ptr: *mut u8, new_size: usize) -> *mut u8 {
+    let ptr = ptr.offset(-(ALIGN as isize));
+    let size = *(ptr as *mut usize);
+    let layout = std::alloc::Layout::from_size_align_unchecked(size + ALIGN, ALIGN);
+
+    let ptr = std::alloc::realloc(ptr, layout, new_size + ALIGN);
+    if ptr.is_null() {
+        return null_mut();
+    }
+
+    *(ptr as *mut usize) = new_size;
+    ptr.offset(ALIGN as isize)
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn wasm_vfs_open(
+    _arg1: *mut sqlite3_vfs,
+    _zName: *const c_char,
+    _arg2: *mut sqlite3_file,
+    _flags: c_int,
+    _pOutFlags: *mut c_int,
+) -> c_int {
+    SQLITE_IOERR
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn wasm_vfs_delete(
+    _arg1: *mut sqlite3_vfs,
+    _zName: *const c_char,
+    _syncDir: c_int,
+) -> c_int {
+    SQLITE_IOERR
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn wasm_vfs_access(
+    _arg1: *mut sqlite3_vfs,
+    _zName: *const c_char,
+    _flags: c_int,
+    _pResOut: *mut c_int,
+) -> c_int {
+    SQLITE_IOERR
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn wasm_vfs_fullpathname(
+    _arg1: *mut sqlite3_vfs,
+    _zName: *const c_char,
+    _nOut: c_int,
+    _zOut: *mut c_char,
+) -> c_int {
+    SQLITE_IOERR
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn wasm_vfs_dlopen(
+    _arg1: *mut sqlite3_vfs,
+    _zFilename: *const c_char,
+) -> *mut c_void {
+    null_mut()
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn wasm_vfs_dlerror(
+    _arg1: *mut sqlite3_vfs,
+    _nByte: c_int,
+    _zErrMsg: *mut c_char,
+) {
+    // no-op
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn wasm_vfs_dlsym(
+    _arg1: *mut sqlite3_vfs,
+    _arg2: *mut c_void,
+    _zSymbol: *const c_char,
+) -> ::std::option::Option<unsafe extern "C" fn(*mut sqlite3_vfs, *mut c_void, *const i8)> {
+    None
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn wasm_vfs_dlclose(_arg1: *mut sqlite3_vfs, _arg2: *mut c_void) {
+    // no-op
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn wasm_vfs_sleep(_arg1: *mut sqlite3_vfs, microseconds: c_int) -> c_int {
+    let target_date = js_sys::Date::now() + (microseconds as f64 / 1000.0);
+    while js_sys::Date::now() < target_date {}
```

Comment:
If this PR supports being built with the `atomics` WASM feature (and people are intending to), maybe allow this to be optimized with [memory_atomic_wait32](https://doc.rust-lang.org/core/arch/wasm32/fn.memory_atomic_wait32.html) instead of a spinloop? This is what the [standard library does](https://github.com/rust-lang/rust/blob/fd8d6fbe505ecf913f5e2ca590c69a7da2789879/library/std/src/sys/pal/wasm/atomics/futex.rs#L22) when [built with atomics support](https://github.com/rust-lang/rust/blob/master/library/std/src/sys/sync/mutex/mod.rs#L9) too.

Reply:
This is a good suggestion, thanks.

Reply:
It's worth noting that atomic waits don't work on the browser's main thread.